A better eslint rules config

finscn opened the pull request:
A better eslint rules config    (maybe).


需要在 package.json里加入下列依赖项: 
```
  "devDependencies": {
    "@eslint-stylistic/metadata": "^1.4.1",
    "@stylistic/eslint-plugin": "^1.4.1",
    "@stylistic/eslint-plugin-migrate": "^1.4.1",
    "@typescript-eslint/eslint-plugin": "^6.12.0",
    "@typescript-eslint/parser": "^6.12.0",
    "eslint": "^8.54.0",
    "eslint-plugin-import": "^2.29.0"
  },
```
Re: #

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

<!-- Note: Makes sure these boxes are checked before submitting your PR - thank you!
- [ ] Your pull request title is using English, it's precise and appropriate.
- [ ] If your pull request has gone "stale", you should **rebase** your work on top of the latest version of the upstream branch.
- [ ] If your commit history is full of small, unimportant commits (such as "fix pep8" or "update tests"), **squash** your commits down to a few, or one, discreet changesets before submitting a pull request.
- [ ] Document new code with comments in source code based on API docs
- [ ] Make sure any runtime log information in `log` , `error` or `new Error('')` has been moved into `EngineErrorMap.md` with an ID, and use `logID(id)` or `new Error(getError(id))` instead.
- To official teams:
  - [ ] Check that your PR is following our [guides](https://github.com/cocos/3d-tasks/blob/master/workflows/readme.md)
-->

finscn committed:
Update .eslintrc.yaml

A better eslint rules config    (maybe).

finscn committed:
Update .eslintrc.yaml

优化规则

finscn committed:
Update .eslintrc.yaml

优化规则

finscn committed:
Update .eslintrc.yaml

优化规则

finscn committed:
Update .eslintrc.yaml

finscn committed:
Update .eslintrc.yaml

finscn committed:
Update .eslintrc.yaml

shrinktofit commented on .eslintrc.yaml:
```diff
@@ -163,3 +149,78 @@ rules:
+    '@typescript-eslint/no-empty-function': off
+    no-multi-spaces: [warn, {'ignoreEOLComments': true, 'exceptions': {'Property': true, 'VariableDeclarator': false } }]
+
+    comma-dangle: [error, {
```
Why don't you choose to disallow trailing commas? I think it'd better to allow. For example, if you had the following array:

```ts
const array = [
  a,
  b
];
```

In this time, you want to append a new item, the diff might be look like:

```diff
- b
+b,
+c
```

In compare with the allowing case, the diff would be:

```diff
  b,
+c,
```

finscn replied:
I've changed the config. I think you are right.

finscn committed:
Update .eslintrc.yaml

finscn committed:
Update .eslintrc.yaml

finscn committed:
Update .eslintrc.yaml

shrinktofit commented:
This PR needs to be refined for following reasons:

1. The PR refers too many lint rule changes, which is hard to be reviewed.

2. To convince our developers, it's required that "why should it be this" reason is documented for each lint rule. For example, we've had:
  
  ![image](https://github.com/cocos/cocos-engine/assets/28807867/e1ef014f-c61f-4e1f-ae35-84e4694b3067)


3. To not bother later PRs, PRs issuing lint changes should also issues the changes applied to related files.

finscn committed:
Update .eslintrc.yaml

finscn committed:
Update .eslintrc.yaml

finscn committed:
Update .eslintrc.yaml

finscn committed:
Update .eslintrc.yaml

finscn committed:
Update .eslintrc.yaml

finscn commented:
> This PR needs to be refined for following reasons:
> 
> 1. The PR refers too many lint rule changes, which is hard to be reviewed.
> 2. To convince our developers, it's required that "why should it be this" reason is documented for each lint rule. For example, we've had:
> 
> ![image](https://user-images.githubusercontent.com/28807867/264549375-e1ef014f-c61f-4e1f-ae35-84e4694b3067.png)
> 
> 3. To not bother later PRs, PRs issuing lint changes should also issues the changes applied to related files.

大部分lint的修改 难以通过一两句注释说明, 而且我的英文不好.
  我非常愿意通过远程会议等方式 和你们做进一步沟通.

finscn committed:
Update .eslintrc.yaml

finscn committed:
Update .eslintrc.yaml

finscn committed:
Update .eslintrc.yaml

finscn committed:
Update .eslintrc.yaml

finscn committed:
Update .eslintrc.yaml

finscn committed:
Update .eslintrc.yaml

finscn committed:
use `@typescript-eslint/indent`

finscn committed:
Update .eslintrc.yaml

'indent' 和 '@typescript-eslint/indent' 要分别定义 ,不能互相代替.

finscn committed:
停用 prefer-template

字符串模板性能低, 总是强制使用未必合适.

dumganhar commented on .eslintrc.yaml:
```diff
@@ -163,3 +160,100 @@ rules:
+    # prefer-const: error
+    # prefer-numeric-literals: off
+
+    prefer-template: off
```
Why to disable `prefer-template`?

finscn replied:
字符串模板性能更低,  开启该功能会强制使用字符串模板, 即使简单的一次拼接(如 `"Hello " + user.name`) 也会被转换成 `'Hello ${user.name}'`. 得不偿失.

dumganhar commented on .eslintrc.yaml:
```diff
@@ -163,3 +160,100 @@ rules:
+    # prefer-numeric-literals: off
+
+    prefer-template: off
+    '@typescript-eslint/no-unsafe-argument': warn
```
This change the default `error` to `warn`. Any reasons for this?

dumganhar replied:
Refer to https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/no-unsafe-argument.md

I think marking it as `error` is better.

finscn replied:
因为按照目前的cocos代码, 如果是 error, 那么会大量标红.如果cocos愿意大规模重构代码, error当然是更好的

dumganhar commented on .eslintrc.yaml:
```diff
@@ -163,3 +160,100 @@ rules:
+    '@typescript-eslint/no-unsafe-argument': warn
+
+    max-len: [warn, 250] # more lenient on max length per line
+    no-console: warn # prefer the uniform logging methods
```
Code in engine should not use `console.log/warn/error/info`  directly.  `error` is better.

finscn replied:
理论上是这样, 但是 `console.log` 无法避免. 无论是开发者的代码 还是引擎代码. 总是会在某些时候需要在控制台打印一些信息. 即使生产环境也是如此.

dumganhar commented on .eslintrc.yaml:
```diff
@@ -163,3 +160,100 @@ rules:
+        'exports': never,
+        'functions': never
+    }]
+    quote-props: [error, consistent-as-needed]
```
This config will make 
```ts
const object1 = {
    foo: 'bar',
    baz: 42,
    'qux-lorem': true, // string with `-` could not be a key without quote.
};
```

to be changed to

```ts
const object1 = {
    'foo': 'bar', // quote ' was added for `foo` property
    'baz': 42,  // quote ' was added for `baz` property
    'qux-lorem': true,
};
```

So I think keep the default behavior is better.

finscn replied:
在实际代码中, 一致性也是很重要的. 在一个json/object的代码块中, key的形式应该尽量保证一致.
 而且在现代编辑器中 有没有引号 会导致 key的颜色不同.  如下图:

<img width="317" alt="image" src="https://github.com/cocos/cocos-engine/assets/288367/22277a32-0a02-45bc-9f70-ecc27a73a60a">

如果实际代码中出现这种 key的颜色不一致的情况, 其实是对阅读代码产生干扰的.

dumganhar commented on .eslintrc.yaml:
```diff
@@ -163,3 +160,100 @@ rules:
+        'functions': never
+    }]
+    quote-props: [error, consistent-as-needed]
+    function-paren-newline: off
```
Its default value is `multiline` which i think it's better. 

If we disable this config, the following code
```ts
function foo (
    bar, baz
) {}

const foo1 = function (
    bar, baz
) {};

const foo2 = (
    bar, baz
) => {};
```

will `not` be formated to 

```ts
function foo (bar, baz) {}

const foo1 = function (bar, baz) {};

const foo2 = (bar, baz) => {};
```

finscn replied:
单行 function 就和 单行 if  单行 for 一样,  除了 "节省行数"  并没有任何其他好处. 反而破坏代码的一致性. 而且会造成一定的阅读困难.
要不要把函数写成一行 应该是开发者根据函数复杂度和代码长度自己决定.
打开这个选项, 会出现很多 "虽然变为了单行, 但是单行长度很长" 的情况. 并不便于代码的阅读.

我个人完全无法接受   单行函数,  单行if  单行for .

dumganhar commented on .eslintrc.yaml:
```diff
@@ -163,3 +160,100 @@ rules:
+    }]
+    quote-props: [error, consistent-as-needed]
+    function-paren-newline: off
+    space-unary-ops: error
```
This rule `space-unary-ops: error` is enabled by default. Why do we have to set it explicitly？

finscn replied:
实际测试下来 ,  space-unary-ops 并不是error,  可能是被继承的配置篡改了.
```
extends:
    - eslint:recommended
    - airbnb-base
```
其实我建议不要继承这些所谓的 "已有的最佳实践", 这样的话 eslint的配置就是黑盒的了.
可以去看看     eslint:recommended 和  airbnb-base里是怎么设置的 适合cocos的就copy过来. 不适合的就删除或者改掉.

dumganhar replied:
I agree with this.

dumganhar commented on .eslintrc.yaml:
```diff
@@ -163,3 +160,100 @@ rules:
+            'afterColon': true,
+            'align': colon
+        }
+    }]
```
This rule is not good for our code. If we add a longer property in object, all previous properties will change.
For instance, if we get
```ts
const obj = {
    hello: 'world',
    a    : 'bbb',
};
```

If we want to add a new property called `foooooooo` like:

```ts
const obj = {
    hello    : 'world',  // code changed
    a        : 'bbb',      // code changed
    foooooooo: 'ahaha',
};
```
We just added a new property, but the whole object definition code were changed.

finscn replied:
嗯 这个我可以理解.按你说的问题不大.

dumganhar commented on .eslintrc.yaml:
```diff
@@ -163,3 +160,100 @@ rules:
+        }
+    }]
+
+    func-names: off
```
The default value of `func-names` is `warn`. Why do we need to disable the warning?
Anonymous functions will hard for debugging.

finscn replied:
打开这个开关, 对于  `arr.forEach( (item, index) => {})`  还有很多 callback/ hook 函数 也要求写名字 太讨厌了.

dumganhar commented on .eslintrc.yaml:
```diff
@@ -163,3 +160,100 @@ rules:
+    }]
+
+    func-names: off
+    default-param-last: off
```
https://eslint.org/docs/latest/rules/default-param-last

It's better to put default argument to the last. Why should we disable this check?

finscn replied:
在实际开发中我经常有类似: 
function   foo(a:number , b:number =1,  c?:string) {  ... }  这样的需求.
就是  可选参数在最后一个.  强制给可选参数赋默认值, 会破坏原有逻辑.   `不传参`和`传默认值` 是完全不同的概念

dumganhar replied:
Got it.

finscn committed:
更新 ignoredNodes

经测试 indent.ignoredNodes 在 eslint标准规则里不能省略.

finscn committed:
新增 member-delimiter-style 

确保 interface 中 声明成员变量后以分号结尾.

finscn committed:
Update .eslintrc.yaml

用  ts plugin来处理 semi,  解决 eslint里 semi 的冲突问题.

finscn committed:
Update .eslintrc.yaml

纠正上一个 commit 里的错误.

finscn committed:
使用 @stylistic

finscn committed:
关闭 no-prototype-builtins

在 90+%的 场景下, 打开此开关 只会带来更繁杂的代码和操作.
且 此开关并不能解决很多根本问题.
比如   obj+....   , String(obj)  Number(obj) .... 所以 对于 原始object, 需要开发者自己去做处理即可.

finscn committed:
完善 comma-dangle 在ts里的设定

finscn committed:
完善规则

finscn committed:
补充     no-case-declarations: error

finscn committed:
Update .eslintrc.yaml

finscn committed:
Update .eslintrc.yaml

finscn committed:
Update .eslintrc.yaml

finscn committed:
Update .eslintrc.yaml

finscn committed:
Update .eslintrc.yaml

finscn committed:
新增 comma-spacing 规则

finscn committed:
完善一些关于空格的规则

finscn committed:
Update .eslintrc.yaml

finscn committed:
新增 function-call-spacing

finscn committed:
关闭 no-inner-declarations 规则

在内部声明函数 是一种常规操作. 没必要禁止.

minggo commented:
一下子改动太多的规则了，会导致引擎出现很多的错误。关于规则的改动得一个个来。

finscn commented:
> 一下子改动太多的规则了，会导致引擎出现很多的错误。关于规则的改动得一个个来。

但是实际情况是 这么久了, 是一个都没改过啊....

minggo commented:
我的意思是你 PR 的时候不要一下子改太多。最好一次一个。这样每个配置可以单独讨论，不会耦合在一起。
